File: src/logic_interpreter/logic_runner.ts
// logic_runner.ts - Global logic runner manager
type LogFunc = (title: string, msg: string) => void;

interface RunningLogic {
    logicId: string;
    stock: string;
    worker: Worker;
    startTime: number;
    interval: number;
}

class LogicRunnerManager {
    private runningLogics: Map<string, RunningLogic> = new Map();
    private logCallbacks: Map<string, LogFunc> = new Map();

    /**
     * 로직을 Worker isolation에서 반복 실행 시작
     */
    public startLogic(
        logicId: string,
        stock: string,
        logicData: any,
        logFunc: LogFunc,
        logDetails: boolean = false,
        interval: number = 5000
    ): boolean {
        // 이미 실행 중인 로직인지 확인
        if (this.runningLogics.has(logicId)) {
            logFunc("Error", `로직 ID "${logicId}"는 이미 실행 중입니다.`);
            return false;
        }

        try {
            // Worker 생성
            const worker = new Worker(
                new URL('./interpreter.worker.ts', import.meta.url),
                { type: 'module' }
            );

            // 로그 콜백 저장
            this.logCallbacks.set(logicId, logFunc);

            // Worker 메시지 핸들러 설정
            worker.onmessage = (e: MessageEvent) => {
                this.handleWorkerMessage(logicId, e.data);
            };

            worker.onerror = (error: ErrorEvent) => {
                const callback = this.logCallbacks.get(logicId);
                if (callback) {
                    callback("Error", `Worker 오류: ${error.message}`);
                }
                this.stopLogic(logicId);
            };

            // 실행 정보 저장
            this.runningLogics.set(logicId, {
                logicId,
                stock,
                worker,
                startTime: Date.now(),
                interval
            });

            // Worker 초기화 및 시작
            worker.postMessage({
                type: 'init',
                stock,
                logicData
            });

            // ready 메시지를 받은 후 start
            const readyHandler = (e: MessageEvent) => {
                if (e.data.type === 'ready') {
                    worker.postMessage({
                        type: 'start',
                        interval,
                        logDetails
                    });
                    worker.removeEventListener('message', readyHandler);
                }
            };
            worker.addEventListener('message', readyHandler);

            logFunc("System", `로직 "${logicId}" 실행 시작 (${interval}ms 간격)`);
            return true;

        } catch (error: any) {
            logFunc("Error", `로직 시작 실패: ${error.message}`);
            return false;
        }
    }

    /**
     * 실행 중인 로직 중지
     */
    public stopLogic(logicId: string): boolean {
        const running = this.runningLogics.get(logicId);
        if (!running) {
            return false;
        }

        // Worker에 중지 메시지 전송
        running.worker.postMessage({ type: 'stop' });
        
        // Worker 종료
        running.worker.terminate();

        // 정리
        this.runningLogics.delete(logicId);
        const callback = this.logCallbacks.get(logicId);
        if (callback) {
            callback("System", `로직 "${logicId}" 실행 중지`);
        }
        this.logCallbacks.delete(logicId);

        return true;
    }

    /**
     * 모든 실행 중인 로직 중지
     */
    public stopAllLogics(): void {
        const logicIds = Array.from(this.runningLogics.keys());
        logicIds.forEach(id => this.stopLogic(id));
    }

    /**
     * 실행 중인 로직 정보 조회
     */
    public getRunningLogic(logicId: string): RunningLogic | undefined {
        return this.runningLogics.get(logicId);
    }

    /**
     * 모든 실행 중인 로직 목록 조회
     */
    public getAllRunningLogics(): RunningLogic[] {
        return Array.from(this.runningLogics.values());
    }

    /**
     * 로직이 실행 중인지 확인
     */
    public isRunning(logicId: string): boolean {
        return this.runningLogics.has(logicId);
    }

    /**
     * Worker로부터 받은 메시지 처리
     */
    private handleWorkerMessage(logicId: string, data: any) {
        const callback = this.logCallbacks.get(logicId);
        const running = this.runningLogics.get(logicId);

        switch (data.type) {
            case 'log':
                if (callback) {
                    callback(data.title, data.msg);
                }
                break;

            case 'order':
                // 주문 처리 (메인 스레드에서 실행)
                if (callback) {
                    this.handleOrder(logicId, data, callback);
                }
                break;

            case 'api-request':
                // Worker에서 온 API 요청 처리
                if (running) {
                    this.handleAPIRequest(running.worker, data);
                }
                break;

            case 'rl-request':
                // Worker에서 온 RL 요청 처리
                if (running) {
                    this.handleRLRequest(running.worker, data);
                }
                break;

            case 'started':
                if (callback) {
                    callback("System", `로직 실행 시작됨 (${data.interval}ms 간격)`);
                }
                break;

            case 'stopped':
                if (callback) {
                    callback("System", "로직 실행 중지됨");
                }
                break;

            case 'error':
                if (callback) {
                    callback("Error", data.message);
                }
                break;

            default:
                console.warn('Unknown worker message type:', data.type);
        }
    }

    /**
     * Worker에서 온 API 요청 처리
     */
    private async handleAPIRequest(worker: Worker, data: any) {
        const { requestId, method, params } = data;

        try {
            const electronAPI = (window as any).electronAPI;
            if (!electronAPI) {
                throw new Error('electronAPI is not available');
            }

            let result: any;
            switch (method) {
                case 'fetchCandles':
                    result = await electronAPI.fetchCandles(...params);
                    break;
                case 'getHighestPrice':
                    result = await electronAPI.getHighestPrice(...params);
                    break;
                default:
                    throw new Error(`Unknown API method: ${method}`);
            }

            // 성공 응답 전송
            worker.postMessage({
                type: 'api-response',
                requestId,
                success: true,
                result
            });
        } catch (error: any) {
            // 에러 응답 전송
            worker.postMessage({
                type: 'api-response',
                requestId,
                success: false,
                error: error.message
            });
        }
    }

    /**
     * Worker에서 온 RL 요청 처리
     */
    private async handleRLRequest(worker: Worker, data: any) {
        const { requestId, data: rlData } = data;

        try {
            // 메인 스레드의 RLConnection을 가져와서 요청 전송
            const { getGlobalRLConnection } = await import('../communicator/RLConnection');
            const rlConnection = getGlobalRLConnection();
            
            if (!rlConnection) {
                throw new Error('RLConnection is not initialized');
            }

            const result = await rlConnection.send(rlData);

            // 성공 응답 전송
            worker.postMessage({
                type: 'rl-response',
                requestId,
                success: true,
                result
            });
        } catch (error: any) {
            // 에러 응답 전송
            worker.postMessage({
                type: 'rl-response',
                requestId,
                success: false,
                error: error.message
            });
        }
    }

    /**
     * Worker에서 요청한 주문 처리
     */
    private async handleOrder(_logicId: string, orderData: any, logFunc: LogFunc) {
        const { action, orderType, stock, limitPrice, quantity } = orderData;
        
        try {
            if (action === 'buy') {
                if (orderType === 'market') {
                    logFunc("Buy", `시장가 ${quantity}₩어치 매수를 시도합니다.`);
                    const result = await (window as any).electronAPI.marketBuy(stock, quantity);
                    if (result.success) {
                        logFunc("Buy", `시장가 ${quantity}₩어치 매수 완료`);
                    } else {
                        logFunc("Error", `시장가 매수 실패: ${result.error}`);
                    }
                } else {
                    logFunc("Buy", `지정가 ${limitPrice}₩에 ${quantity}₩어치 매수를 시도합니다.`);
                    const result = await (window as any).electronAPI.limitBuyWithKRW(stock, limitPrice, quantity);
                    if (result.success) {
                        logFunc("Buy", `지정가 ${limitPrice}₩에 ${quantity}₩어치 매수 완료`);
                    } else {
                        logFunc("Error", `지정가 매수 실패: ${result.error}`);
                    }
                }
            } else if (action === 'sell') {
                if (orderType === 'market') {
                    logFunc("Sell", `시장가 ${quantity}₩어치 매도를 시도합니다.`);
                    const result = await (window as any).electronAPI.marketSell(stock, quantity);
                    if (result.success) {
                        logFunc("Sell", `시장가 ${quantity}₩어치 매도 완료`);
                    } else {
                        logFunc("Error", `시장가 매도 실패: ${result.error}`);
                    }
                } else {
                    logFunc("Sell", `지정가 ${limitPrice}₩에 ${quantity}₩어치 매도 시도`);
                    const result = await (window as any).electronAPI.limitSellWithKRW(stock, limitPrice, quantity);
                    if (result.success) {
                        logFunc("Sell", `지정가 ${limitPrice}₩에 ${quantity}₩어치 매도 완료`);
                    } else {
                        logFunc("Error", `지정가 매도 실패: ${result.error}`);
                    }
                }
            }
        } catch (error: any) {
            logFunc("Error", `주문 처리 중 오류: ${error.message}`);
        }
    }
}

// 전역 싱글톤 인스턴스
export const logicRunner = new LogicRunnerManager();

/**
 * 로직 실행 함수 (기존 runLogic 대체)
 */
export function runLogic(
    stock: string,
    logicData: any,
    logFunc: LogFunc,
    logDetails: boolean = false,
    logicId?: string,
    interval: number = 5000
) {
    const id = logicId || `logic-${Date.now()}`;
    return logicRunner.startLogic(id, stock, logicData, logFunc, logDetails, interval);
}

/**
 * 로직 중지 함수
 */
export function stopLogic(logicId: string) {
    return logicRunner.stopLogic(logicId);
}

/**
 * 모든 로직 중지 함수
 */
export function stopAllLogics() {
    logicRunner.stopAllLogics();
}

/**
 * 실행 중인 로직 정보 조회
 */
export function getRunningLogicInfo(logicId: string) {
    return logicRunner.getRunningLogic(logicId);
}

/**
 * 모든 실행 중인 로직 조회
 */
export function getAllRunningLogics() {
    return logicRunner.getAllRunningLogics();
}

/**
 * 로직 실행 중 여부 확인
 */
export function isLogicRunning(logicId: string) {
    return logicRunner.isRunning(logicId);
}

File: backend/logic_runner.py
"""
Logic Runner Manager
여러 로직을 동시에 실행하고 관리하는 매니저
"""
import asyncio
from typing import Dict, Optional, Callable, Any
from datetime import datetime
from interpreter import Interpreter


class RunningLogic:
    """실행 중인 로직 정보"""
    
    def __init__(
        self, 
        logic_id: str, 
        stock: str, 
        interpreter: Interpreter,
        interval: float,
        task: asyncio.Task
    ):
        self.logic_id = logic_id
        self.stock = stock
        self.interpreter = interpreter
        self.interval = interval
        self.task = task
        self.start_time = datetime.now()
        self.is_running = True
    
    def stop(self):
        """로직 실행 중지"""
        self.is_running = False
        if not self.task.cancelled():
            self.task.cancel()


class LogicRunner:
    """로직 실행 관리자"""
    
    def __init__(self, upbit_api):
        self.upbit_api = upbit_api
        self.running_logics: Dict[str, RunningLogic] = {}
        self.log_callbacks: Dict[str, Callable[[str, str], None]] = {}
    
    async def start_logic(
        self,
        logic_id: str,
        stock: str,
        logic_data: Dict[str, Any],
        log_func: Callable[[str, str], None],
        log_details: bool = False,
        interval: float = 5.0  # seconds
    ) -> bool:
        """
        로직 실행 시작
        
        Args:
            logic_id: 로직 고유 ID
            stock: 거래 종목 (예: KRW-BTC)
            logic_data: 로직 데이터 (buyGraph, sellGraph 포함)
            log_func: 로그 콜백 함수
            log_details: 상세 로그 여부
            interval: 실행 간격 (초)
        
        Returns:
            성공 여부
        """
        # 이미 실행 중인지 확인
        if logic_id in self.running_logics:
            log_func("Error", f"로직 ID '{logic_id}'는 이미 실행 중입니다.")
            return False
        
        try:
            # Interpreter 생성 및 초기화
            interpreter = Interpreter(self.upbit_api, log_func)
            interpreter.set_stock(stock)
            interpreter.parse(logic_data)
            
            if not interpreter.parse_complete:
                log_func("Error", "로직 파싱에 실패했습니다.")
                return False
            
            # 데이터 매니저 초기화
            await interpreter.initialize()
            
            # 로그 콜백 저장
            self.log_callbacks[logic_id] = log_func
            
            # 실행 태스크 생성
            task = asyncio.create_task(
                self._run_logic_loop(logic_id, interpreter, log_details, interval)
            )
            
            # 실행 정보 저장
            running_logic = RunningLogic(logic_id, stock, interpreter, interval, task)
            self.running_logics[logic_id] = running_logic
            
            log_func("System", f"로직 '{logic_id}' 실행 시작 ({interval}초 간격)")
            return True
        
        except Exception as e:
            log_func("Error", f"로직 시작 실패: {str(e)}")
            return False
    
    async def _run_logic_loop(
        self,
        logic_id: str,
        interpreter: Interpreter,
        log_details: bool,
        interval: float
    ):
        """로직 실행 루프"""
        running_logic = self.running_logics.get(logic_id)
        log_func = self.log_callbacks.get(logic_id)
        
        try:
            while running_logic and running_logic.is_running:
                try:
                    # 로직 실행
                    await interpreter.run(log_details)
                    
                    # 다음 실행까지 대기
                    await asyncio.sleep(interval)
                
                except asyncio.CancelledError:
                    if log_func:
                        log_func("System", f"로직 '{logic_id}' 실행 취소됨")
                    break
                
                except Exception as e:
                    if log_func:
                        log_func("Error", f"로직 실행 중 오류: {str(e)}")
                    # 오류 발생 시에도 계속 실행
                    await asyncio.sleep(interval)
        
        finally:
            # 정리
            if log_func:
                log_func("System", f"로직 '{logic_id}' 실행 종료")
    
    def stop_logic(self, logic_id: str) -> bool:
        """
        로직 실행 중지
        
        Args:
            logic_id: 중지할 로직 ID
        
        Returns:
            성공 여부
        """
        running_logic = self.running_logics.get(logic_id)
        
        if not running_logic:
            return False
        
        # 로직 중지
        running_logic.stop()
        
        # 정리
        del self.running_logics[logic_id]
        
        log_func = self.log_callbacks.get(logic_id)
        if log_func:
            log_func("System", f"로직 '{logic_id}' 실행 중지")
            del self.log_callbacks[logic_id]
        
        return True
    
    def stop_all_logics(self):
        """모든 로직 중지"""
        logic_ids = list(self.running_logics.keys())
        for logic_id in logic_ids:
            self.stop_logic(logic_id)
    
    def get_running_logic(self, logic_id: str) -> Optional[Dict[str, Any]]:
        """실행 중인 로직 정보 조회"""
        running_logic = self.running_logics.get(logic_id)
        
        if not running_logic:
            return None
        
        return {
            'logicId': running_logic.logic_id,
            'stock': running_logic.stock,
            'interval': running_logic.interval,
            'startTime': running_logic.start_time.isoformat(),
            'isRunning': running_logic.is_running
        }
    
    def get_all_running_logics(self) -> list:
        """모든 실행 중인 로직 목록 조회"""
        return [
            self.get_running_logic(logic_id)
            for logic_id in self.running_logics.keys()
        ]
    
    def is_running(self, logic_id: str) -> bool:
        """로직 실행 중 여부 확인"""
        return logic_id in self.running_logics
